internal/app/machined/internal/server/v1alpha1/v1alpha1_server.go

```diff
@@ -422,14 +424,81 @@ func (s *Server) Upgrade(ctx context.Context, in *machine.UpgradeRequest) (reply
 	return reply, nil
 }
 
+// ResetOptions implements runtime.ResetOptions interface.
+type ResetOptions struct {
+	*machine.ResetRequest
+
+	systemDiskTargets []*installer.Target
+}
+
+// GetSystemDiskTargets implements runtime.ResetOptions interface.
+func (opt *ResetOptions) GetSystemDiskTargets() []runtime.DiskTarget {
+	if opt.systemDiskTargets == nil {
+		return nil
+	}
+
+	result := make([]runtime.DiskTarget, len(opt.systemDiskTargets))
+
+	for i := range result {
+		result[i] = opt.systemDiskTargets[i]
+	}
+
+	return result
+}
+
 // Reset resets the node.
 //
-// nolint: dupl
+// nolint: dupl, gocyclo
 func (s *Server) Reset(ctx context.Context, in *machine.ResetRequest) (reply *machine.ResetResponse, err error) {
 	log.Printf("reset request received")
 
+	opts := ResetOptions{
+		ResetRequest: in,
+	}
+
+	if len(in.GetSystemDisk()) > 0 {
```

Comment:
verifying options early in the API flow to return errors and not start the reset sequence if it's set to fail